### ISSUE TITLE
One short NAS Supernet training strategy

### ISSUE THREAD

**Issue by CanyonWind** [megvii-model/ShuffleNet-Series#13]
Hi, I've implemented an [MXNet version](https://github.com/CanyonWind/oneshot_nas), with both fixed structure model and block & channel selection supernet model, based on this official release. 

But others and myself all have had a hard time to train the supernet with both block and channel selection from scratch. Currently, my work-around is to only train with the block selection at the first 60 epochs and the train with both block and channel selection for the rest 60. 

Could you please let me know what's your training strategy for the supernet? Did you train block and channel selection from scratch? 

Thanks

**Comment by CanyonWind** [comment]
Besides whether to do block and channel selection from scratch, another problem I met in supernet training is that, according to my experiments, **channel selection seems only working with the candidate range (0.6 ~ 2.0) * base_mid_channels, but not the smaller scales like 0.2 or 0.4**.

In the official code, you have the channel candidate scales in between [0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6] and selected one of them multiplying base_mid_channel as the # mid_channels for each block. 

The same philosophy was followed and I searched for range [0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0]. To be specific, **in the first 60 epochs** I only did training with block selection and the mid_channel was set to be the maximum [2.0] * base_mid_channel. It converged fine. **Start from epoch 61**, the supernet was gradually allowed to select the mid channels from [2.0] to [1.8, 2.0], [1.6, 1.8, 2.0] ... [0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] * base_mid_channels.

What I found confusing is that Channel Selection worked okay from [2.0] to [0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] but after that, when training with smaller channel scales like 0.4 and 0.2, it crashed with free-fall dropping accuracy.

What do you think might be wrong? And how did you utilize the uniform random choices in the training stage, did you apply different block choice and channel choice for each batch ([like this](https://github.com/CanyonWind/oneshot_nas/blob/43432c5e1677df8c43f432afeffdfbfd28d20a07/train_imagenet.py#L443-L452)) or each epoch?

Any feedback would be much appreciated.

**Comment by CanyonWind** [comment]
Two schedules for channel selection warm-up were tried:
``` python
   # First schedule: increase one channel choice per epoch after the 60th epoch
   1 - 60 epochs:          Only block selection (BS)
   61 epoch:               [1.8, 2.0] + BS
   62 epoch:               [1.6, 1.8, 2.0] + BS
   63 epoch:               [1.4, 1.6, 1.8, 2.0] + BS
   64 epoch:               [1.2, 1.4, 1.6, 1.8, 2.0] + BS
   65 epoch:               [1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS
   66 epoch:               [0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS
   67 epoch:               [0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS 
   68 epoch:               [0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS 
   69 epoch:               [0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS
```
``` python
   # Second schedule: apply more epochs for [0.2, 0.4, 0.6, 0.8, 1.0] channel choices
   1 - 60 epochs:          Only block selection (BS)
   61 epoch:               [1.8, 2.0] + BS
   62 epoch:               [1.6, 1.8, 2.0] + BS
   63 epoch:               [1.4, 1.6, 1.8, 2.0] + BS
   64 epoch:               [1.2, 1.4, 1.6, 1.8, 2.0] + BS
   65 - 66 epochs:         [1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS
   67 - 69 epochs:         [0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS
   70 - 73 epochs:         [0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS 
   74 - 78 epochs:         [0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS 
   79 - 120 epochs:        [0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6, 1.8, 2.0] + BS
```
They both failed (training accuracy drops to 1/1000) when the supernet is allowed to select 0.4 or smaller scales.  Also, I realized that the official one-shot nas searched channel scales also does not contain [0.4, 0.2]. Is this phenomenon found in your experiment too? 

```python
   # Official one-shot nas searched channel scales
   scale_list:             [0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.4, 1.6]
   scale_ids:              [6, 5, 3, 5, 2, 6, 3, 4, 2, 5, 7, 5, 4, 6, 7, 4, 4, 5, 4, 3]
```

**Comment by CanyonWind** [comment]
🙇 Anyone to help please  ...

**Comment by xieydd** [comment]
@CanyonWind Hey, I form a group of reproducting  spos, can you join us. You can contact me with mail: xieydd@gmail.com or wechat: 13253537292

**Comment by githubgzc** [comment]
> Anyone to help please...
@CanyonWind 
Hello, our supernet trainging strategy is the same as the final model training strategy. And we didn't use any special schedule for channel selection, just uniform sampling.  Do you use the weight sharing strategy of channel search in your supernet training? 
If you have any other questions, you can contact me with email : guozichao@megvii.com.

**Comment by CanyonWind** [comment]
Hi @xieydd, thanks for the invitation, but I don't understand well what is spos

**Comment by CanyonWind** [comment]
> Hello, our supernet trainging strategy is the same as the final model training strategy. And we didn't use any special schedule for channel selection, just uniform sampling.  Do you use the weight sharing strategy of channel search in your supernet training?

Thanks @githubgzc! I did use weight sharing during supernet training. Although I don't know how you achieved it, the way I implemented was inserting channel masks to deactivate un-selected channels and their gradients per batch. It means that, in my implementation, each batch will be trained with different block choices and channel choices.  

Two quick questions. Did you **train from the first epoch** with channel selection **with all the choices (0.2 ~ 1.6)**? Did you sample different block & channel choices **per batch or per epoch**? 

Thanks again

**Comment by xieydd** [comment]
@CanyonWind single path one-shot NAS, haha.

**Comment by githubgzc** [comment]
> > Hello, our supernet trainging strategy is the same as the final model training strategy. And we didn't use any special schedule for channel selection, just uniform sampling.  Do you use the weight sharing strategy of channel search in your supernet training?
> 
> Thanks @githubgzc! I did use weight sharing during supernet training. Although I don't know how you achieved it, the way I implemented was inserting channel masks to deactivate un-selected channels and their gradients per batch. It means that, in my implementation, each batch will be trained with different block choices and channel choices.
> 
> Two quick questions. Did you **train from the first epoch** with channel selection **with all the choices (0.2 ~ 1.6)**? Did you sample different block & channel choices **per batch or per epoch**?
> 
> Thanks again

We train from the first epoch and sample different channel choices per batch.

**Comment by CanyonWind** [comment]
> @CanyonWind single path one-shot NAS, haha.

I see... lol

**Comment by CanyonWind** [comment]
> single path one-shot NAS, haha.
I see... lol

**Comment by CanyonWind** [comment]
> We train from the first epoch and sample different channel choices per batch.

Thanks. Could you please also share some insights about how did you evaluate validation accuracy among supernet training. Did you also sample random block & channel choices (the combination of which very likely haven't been seen during any training batch) per validation batch?

**Comment by githubgzc** [comment]
> > We train from the first epoch and sample different channel choices per batch.
> 
> Thanks. Could you please also share some insights about how did you evaluate validation accuracy among supernet training. Did you also sample random block & channel choices (the combination of which very likely haven't been seen during any training batch) per validation batch?

It is hard to evaluate validation accuracy correctly among supernet training because there are lots of sub networks. You can randomly sample one sub network and test it with the whole validation dataset.

**Comment by CanyonWind** [comment]
> It is hard to evaluate validation accuracy correctly among supernet training because there are lots of sub networks. You can randomly sample one sub network and test it with the whole validation dataset.

This aligns with my interpretation. For me, I still did random block & channel sampling per batch for the validation set, but, because of the randomness and the val acc fluctuation, it makes me hard to pick the best performing iter of the supernet. I'm wondering how did you pick that?

**Comment by githubgzc** [comment]
We did'nt pick the best performing iter of the supernet, we use the last.

**Comment by CanyonWind** [comment]
Thanks!

**Comment by CanyonWind** [comment]
Sorry to reopen it again. One quick question, did you also update the BN statistics for the sampled (random or not) subnet when verifying the validation accuracy during training? Thanks.

**Comment by githubgzc** [comment]
> Sorry to reopen it again. One quick question, did you also update the BN statistics for the sampled (random or not) subnet when verifying the validation accuracy during training? Thanks.

Yes. The BN statistics of each subnet should be updated before validation.

**Comment by CanyonWind** [comment]
Appreciate it
